tempesta_fw/http.c

```diff
@@ -777,6 +1288,78 @@ tfw_http_adjust_resp(TfwHttpResp *resp, TfwHttpReq *req)
 				     TFW_HTTP_HDR_SERVER, 0);
 }
 
+/*
+ * Forward responses to the client in the correct order.
+ */
+void
+tfw_http_resp_fwd(TfwHttpReq *req, TfwHttpResp *resp)
+{
+	TfwHttpReq *tmp;
+	TfwConnection *cli_conn = req->conn;
+	struct list_head *seq_queue = &cli_conn->msg_queue;
+	LIST_HEAD(out_queue);
+
+	TFW_DBG2("%s: req=[%p], resp=[%p]\n", __func__, req, resp);
+
+	/*
+	 * Starting with the first request on the list, pick consecutive
+	 * requests that have a paired response. Remove those requests
+	 * from the list, and put them on the list of outgoing responses.
+	 *
+	 * However, if the list is empty, then then it's either a bug,
+	 * or the client connection had been closed. If it's a bug, then
+	 * the correct order of responses to requests may be broken. The
+	 * client connection needs to be closed.
+	 */
+	spin_lock(&cli_conn->msg_qlock);
+	if (list_empty(seq_queue)) {
+		spin_unlock(&cli_conn->msg_qlock);
+		TFW_DBG2("%s: The client's request missing: conn=[%p]\n",
+			 __func__, cli_conn);
+		ss_close_sync(cli_conn->sk, true);
+		tfw_http_conn_msg_free((TfwHttpMsg *)resp);
+		tfw_http_conn_msg_free((TfwHttpMsg *)req);
+		return;
+	}
+	req->resp = (TfwHttpMsg *)resp;
+	do {
+		req = list_first_entry(seq_queue, TfwHttpReq, msg.seq_list);
+		if (req->resp == NULL)
+			break;
+		list_move_tail(&req->msg.seq_list, &out_queue);
+	} while (!list_empty(seq_queue));
+	spin_unlock(&cli_conn->msg_qlock);
+
+	/* Forward responses to the client. */
+	list_for_each_entry_safe(req, tmp, &out_queue, msg.seq_list) {
+		list_del_init(&req->msg.seq_list);
+		resp = (TfwHttpResp *)req->resp;
+		/*
+		 * If the client connection is dead, then discard all
+		 * @req and @resp in the @out_queue. Remaining requests
+		 * from the client in the @seq_queue will be handled at
+		 * the time the client connection is released.
+		 */
+		if (!tfw_connection_live(cli_conn)) {
```

Comment:
Indeed. Thanks. Still, a client connection may become dead before/while we attempt to forward a bunch of responses to the connection. The only check on whether the connection is alive is made on the socket level in `ss_do_send()` which is the time the SKB are about to be submitted to Linux TCP/IP stack. However before that happens, the request's data is added to the work queue, the work is scheduled on another CPU, and then data is processed in a separate softirq thread. All of that unnecessarily takes resources. Is that okay? May be it would be useful to use the same `tfw_connection_put_to_death()` on client connections as well?